### PR TITLE
MET-3903: Include datasetId filtering

### PR DESCRIPTION
**Pull request by JoanaCMS**

### CODE REVIEW

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/request/FiltersWrapper.java [hunk 1]
```diff
@@ -10,6 +12,8 @@
 @JsonSerialize
 public class FiltersWrapper {
 
+  @JsonProperty(FilterNames.DATASET_ID)
+  private String datasetId;
```
Comment (stzanakis) [inline statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/request/FiltersWrapper.java:16]: But isn't this a filter too? Why is it separated?
I think we could have it as part of the other filters which makes it more generic and also would support all other functionality that the other filters provide right?
(We could restrict that specific filter in code, to not have breakdowns for example, if we wanted it to but I don't think we need to.)

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java [hunk 2]
```diff
@@ -150,7 +166,17 @@ private StatisticsQuery prepareFilteringQuery(StatisticsFilteringRequest statist
 
   private Map<MongoStatisticsField, ValueRange> prepareRangeFilteringOptionsQuery(StatisticsQuery query) {
     Map<MongoStatisticsField, ValueRange> result = new EnumMap<>(MongoStatisticsField.class);
-    MongoStatisticsField.getRangeFields().forEach(field -> result.put(field, query.queryForValueRange(field)));
+    MongoStatisticsField.getRangeFields().forEach(field ->
+        {
+            ValueRange valueRange;
+            if (query.queryForValueRange(field) == null){
```
Comment (stzanakis) [inline statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java:172]: Why is null not okay?
And if it really isn't okay then perhaps the `new ValueRange("", ""); ` would better happen in the function itself.

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java [hunk 3]
```diff
@@ -92,7 +95,15 @@ public FilteringResult queryDataWithFilters(StatisticsFilteringRequest statistic
     Map<MongoStatisticsField, ValueRange> rangeAvailableOptions = prepareRangeFilteringOptionsQuery(readyQuery);
 
     int totalRecords = queryResult.getRecordCount();
-    StatisticsResult statisticsAllRecordsResult = new StatisticsResult(STATISTICS_RESULT_ROOT_VALUE, totalRecords, 100);
+
+    StatisticsResult statisticsAllRecordsResult;
+    if(totalRecords == 0) {
```
Comment (stzanakis) [inline statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java:100]: I think that the percentage could be "calculated" in the constructor(and setCount).
I mean it is based on a the total record number, it's not a value that we get from somewhere and it's relevant to that specific class.
So I would move that responsibility to the class.
The percentage would not be part of the constructor parameters, nor being allowed to be set(setter) and only allow it to be calculated inside the class (constructor/setCount).
How does that sound?

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/request/FiltersWrapper.java (hunk 1 above)
Comment (JoanaCMS) [reply statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/request/FiltersWrapper.java:16]: After our chat, we agreed on introducing a structurally "non-breakdownable"  filter type. 
I will do the necessary changes

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java (hunk 2 above)
Comment (JoanaCMS) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java:172]: After some clarification, we decided on changing the original method to return `Optional<ValueRange>`. I will add those changes

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java [hunk 4]
```diff
@@ -183,6 +222,9 @@ private StatisticsQuery prepareFilteringQuery(StatisticsFilteringRequest statist
   }
 
   private double calculatePercentage(double totalCount, double count) {
+    if (totalCount <= 0) {
```
Comment (stzanakis) [inline statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java:225]: This is great, but I would intrigue you to try to make this whole method a one liner return :)

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/persistence/StatisticsQuery.java [hunk 5]
```diff
@@ -233,6 +235,11 @@ public ValueRange queryForValueRange(MongoStatisticsField mongoStatisticsField)
     private final String from;
     private final String to;
 
+    public ValueRange(){
```
Comment (stzanakis) [inline statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/persistence/StatisticsQuery.java:238]: I think this is not required anymore

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/request/StatisticsNonBreakdownValueFilter.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.Collections;
+import java.util.List;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * Similarly to filter type {@link StatisticsValueFilter} but this class does not contain breakdowns
+ */
+
+@JsonSerialize
+public class StatisticsNonBreakdownValueFilter implements StatisticsFilter{
```
Comment (stzanakis) [inline statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/request/StatisticsNonBreakdownValueFilter.java:14]: Hmm, I'm thinking. Could we make the hierarchy more streamlined? I mean:
StatisticsFilter -> StatisticsValueFilter(contains only values) -> StatisticsBreakdownValueFilter(now this inherits values and contains breakdown)
Not sure if it's possible in regards to (de)serialization, didn't try it..

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/request/StatisticsFilteringRequest.java [hunk 7]
```diff
@@ -129,6 +133,14 @@ public void setUpdatedDate(StatisticsRangeFilter updatedDate) {
...
 
+  public StatisticsNonBreakdownValueFilter getDatasetId() {
+    return datasetId;
+  }
+
+  public void setDatasetId(StatisticsNonBreakdownValueFilter datasetId) {
+    this.datasetId = datasetId;
+  }
+
   public List<StatisticsValueFilter> getAllValueFilters(){
```
Comment (stzanakis) [inline statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/request/StatisticsFilteringRequest.java:144]: Shouldn't the datasetId be part of this call too?

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java [hunk 8]
```diff
@@ -114,56 +131,77 @@ public FilteringResult queryDataWithFilters(StatisticsFilteringRequest statistic
...
+    List<MongoStatisticsField> breakdowns = RequestUtils.parseBreakdownsFromRequest(filters);
 
     // Set each parsed value into query
     parsedValueFilters.forEach(query::withValueFilter);
-    parsedRangeFilters.forEach((key, value) -> query.withRangeFilter(key, value.getFrom(), value.getTo()));
+    parsedRangeFilters.forEach(
+        (key, value) -> query.withRangeFilter(key, value.getFrom(), value.getTo()));
     query.withBreakdowns(breakdowns.toArray(MongoStatisticsField[]::new));
 
+    List<String> datasetIdValues = statisticsRequest.getFilters().getDatasetId().getValues();
```
Comment (stzanakis) [inline statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java:164]: This should be ideally handled from RequestUtils.parseValuesFiltersFromRequest call. That means that the eu.europeana.statistics.dashboard.common.iternal.MongoStatisticsField#DATASET_ID should be updated accordingly. In that way the code block here can be omitted completely.

Comment (JoanaCMS) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java:164]: I have made the necessary changes and removed this block of code thanks to the previous comment's changes

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/persistence/StatisticsQuery.java (hunk 5 above)
Comment (JoanaCMS) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/persistence/StatisticsQuery.java:238]: Indeed, not necessary anymore. Removed

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java (hunk 4 above)
Comment (JoanaCMS) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java:225]: I did it! :)

**Pull request merged**